subsys/sensing/sensor_mgmt.c

```diff
@@ -0,0 +1,329 @@
+		conn_sample_size += get_sensor_sample_size_from_dt(dt, i);
+	}
+
+	/* total memory to be allocated for a sensor according to sensor device tree:
+	 * 1) size of struct sensing_sensor *
+	 * 2) sample data point to struct sensing_sensor->data_buf
+	 * 3) size of struct sensing_connection* for sensor connection to its reporter
+	 * 4) reporter sample size to be stored in connection data
+	 */
+	total_size = sizeof(*sensor) + sample_size +
+		     dt->reporter_num * sizeof(*sensor->conns) + conn_sample_size;
+
+	/* total size for different sensor maybe different, for example:
+	 * there's no reporter for physical sensor, so no connection memory is needed
+	 * reporter num of each virtual sensor may also different, so connection memory is also
+	 * varied, so here malloc is a must for different sensor.
+	 */
+	sensor = malloc(total_size);
+	if (!sensor) {
+		LOG_ERR("malloc memory for struct sensing_sensor error");
+		return NULL;
+	}
+	sensor->sample_size = sample_size;
+	sensor->data_buf = (uint8_t *)sensor + sizeof(*sensor);
+	sensor->conns = (struct sensing_connection *)((uint8_t *)sensor->data_buf + sample_size);
+
+	tmp_data = sensor->conns + dt->reporter_num;
+	for (i = 0; i < dt->reporter_num; i++) {
+		sensor->conns[i].data = tmp_data;
+		tmp_data = (uint8_t *)tmp_data + get_sensor_sample_size_from_dt(dt, i);
+	}
+
+	__ASSERT(tmp_data == ((uint8_t *)sensor + total_size), "sensor memory assign error");
+
+	sensor->dt = dt;
+	sensor->dt->info.flags = sensor_ctx->register_info->flags;
+	sensor->dt->info.version = sensor_ctx->register_info->version;
+
+#ifdef CONFIG_SENSOR_INFO
+	STRUCT_SECTION_FOREACH(sensor_info, tmp) {
+		if (sensor->dt->dev == tmp->dev) {
+			sensor->dt->info.vendor = tmp->vendor;
+			sensor->dt->info.model = tmp->model;
+			break;
+		}
+	}
+#endif
+
+	LOG_INF("create sensor, sensor:%s, min_ri:%d(us)",
+		sensor->dt->dev->name, sensor->dt->info.minimal_interval);
+
+	sensor->interval = 0;
+	sensor->sensitivity_count = sensor_ctx->register_info->sensitivity_count;
+	__ASSERT(sensor->sensitivity_count <= CONFIG_SENSING_MAX_SENSITIVITY_COUNT,
+		 "sensitivity count:%d should not exceed MAX_SENSITIVITY_COUNT",
+		 sensor->sensitivity_count);
+	memset(sensor->sensitivity, 0x00, sizeof(sensor->sensitivity));
+
+	sensor->state = SENSING_SENSOR_STATE_OFFLINE;
+	sys_slist_init(&sensor->client_list);
+
+	sensor_ctx->priv_ptr = sensor;
+
+	return sensor;
+}
+
+static int sensing_init(void)
+{
+	struct sensing_context *ctx = &sensing_ctx;
+	struct sensing_dt_info **dt_list = sensors_dt_list;
+	struct sensing_sensor *sensor;
+	enum sensing_sensor_state state;
+	int ret = 0;
+	int i;
+
+	LOG_INF("sensing init begin...");
+
+	if (ctx->sensing_initialized) {
+		LOG_INF("sensing is already initialized");
+		return 0;
+	}
+
+	if (ctx->sensor_num == 0) {
+		LOG_WRN("no sensor created by device tree yet");
+		return 0;
+	}
+
+	for (i = 0; i < ctx->sensor_num; i++) {
+		sensor = create_sensor(*dt_list++);
```

Comment:
Needed when you delete SENSING_DT_ITEM macro.
```suggestion
		sensor = create_sensor(dt_list++);
```